node/src/components/consensus/highway_core/active_validator.rs

```diff
@@ -180,21 +185,43 @@ impl<C: Context> ActiveValidator<C> {
         let r_exp = self.round_exp(state, timestamp);
         let r_id = state::round_id(timestamp, r_exp);
         let r_len = state::round_len(r_exp);
-        if timestamp == r_id && state.leader(r_id) == self.vidx {
-            effects.extend(self.request_new_block(state, instance_id, timestamp, rng))
-        } else if timestamp == r_id + self.witness_offset(r_len) {
-            let panorama = self.panorama_at(state, timestamp);
-            if panorama.has_correct() {
-                if let Some(witness_unit) =
-                    self.new_unit(panorama, timestamp, None, state, instance_id, rng)
-                {
-                    effects.push(Effect::NewVertex(ValidVertex(Vertex::Unit(witness_unit))))
+        // Only create new units if enough validators are online.
+        if self.enough_validators_online(state, timestamp) {
+            if timestamp == r_id && state.leader(r_id) == self.vidx {
+                effects.extend(self.request_new_block(state, instance_id, timestamp, rng));
+                return effects;
+            } else if timestamp == r_id + self.witness_offset(r_len) {
+                let panorama = self.panorama_at(state, timestamp);
+                if panorama.has_correct() {
+                    if let Some(witness_unit) =
+                        self.new_unit(panorama, timestamp, None, state, instance_id, rng)
+                    {
+                        effects.push(Effect::NewVertex(ValidVertex(Vertex::Unit(witness_unit))));
+                        return effects;
+                    }
                 }
             }
         }
+        // We are not creating a new unit. Send a ping if necessary, to show that we're online.
+        if !state.has_ping(self.vidx, timestamp) {
+            warn!(%timestamp, "too many validators offline, sending ping");
+            let ping = Ping::new(self.vidx, timestamp, &self.secret, rng);
+            effects.push(Effect::NewVertex(ValidVertex(Vertex::Ping(ping))));
+        }
         effects
     }
 
+    /// Returns whether enough validators are online to finalize values with the target fault
+    /// tolerance threshold, always counting this validator as online.
+    fn enough_validators_online(&self, state: &State<C>, timestamp: Timestamp) -> bool {
+        let target_quorum = (state.total_weight() + self.target_ftt) / 2;
+        let mut weight = state.online_and_honest_weight(timestamp);
+        if !state.is_online_and_honest(self.vidx, timestamp) {
```

Comment:
Do we really have to check if we're online ourselves?

Reply:
No, we don't; that's the point of this line: We should always count ourselves as online, even if we just restarted and the `state` doesn't know about it yet. That's why, if the state doesn't know we're online, we add our own weight anyway.

Reply:
That confused me – that we first go through `state` and return the weight of all honest and online validators but we check if we are in that set ourselves.

Reply:
Good point, I'll just move that weight calculation to `active_validator`.